src/adera/ShipResources.cpp

```diff
@@ -0,0 +1,119 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include "ShipResources.h"
+#include "osp/Active/ActiveScene.h"
+
+using namespace osp;
+using namespace osp::active;
+using namespace adera::active::machines;
+
+/* MachineContainer */
+
+void MachineContainer::propagate_output(WireOutput* output)
+{
+
+}
+
+WireInput* MachineContainer::request_input(WireInPort port)
+{
+    return nullptr;
+}
+
+WireOutput* MachineContainer::request_output(WireOutPort port)
+{
+    return nullptr;
+}
+
+std::vector<WireInput*> MachineContainer::existing_inputs()
+{
+    return m_inputs;
+}
+
+std::vector<WireOutput*> MachineContainer::existing_outputs()
+{
+    return m_outputs;
+}
+
+/* SysMachineContainer */
+
+SysMachineContainer::SysMachineContainer(ActiveScene& rScene)
+    : SysMachine<SysMachineContainer, MachineContainer>(rScene)
+    , m_updateContainers(rScene.get_update_order(), "mach_container", "", "mach_rocket",
+        [this](ActiveScene& rScene) { this->update_containers(rScene); })
+{ }
+
+void SysMachineContainer::update_containers(ActiveScene& rScene)
+{
+    auto view = rScene.get_registry().view<MachineContainer>();
+
+    for (ActiveEnt ent : view)
+    {
+        auto& container = view.get<MachineContainer>(ent);
+
+        // Do something useful here... or not
+    }
+}
+
+float SysMachineContainer::resource_volume(ShipResourceType type, uint64_t quantity)
+{
+    uint64_t units = quantity / (1ull << type.m_quanta);
+    return static_cast<float>(units) * type.m_volume;
+}
+
+float SysMachineContainer::resource_mass(ShipResourceType type, uint64_t quantity)
+{
+    uint64_t units = quantity / (1ull << type.m_quanta);
+    return static_cast<float>(units) * type.m_mass;
+}
+
+uint64_t SysMachineContainer::resource_capacity(ShipResourceType type, float volume)
+{
+    double units = volume / type.m_volume;
+    double quantaPerUnit = static_cast<double>(1ull << type.m_quanta);
+    return static_cast<uint64_t>(units * quantaPerUnit);
+}
+
+Machine& SysMachineContainer::instantiate(ActiveEnt ent,
+    PrototypeMachine config, BlueprintMachine settings)
+{
+    float capacity = std::get<double>(config.m_config["capacity"]);
+    
+    // Currently disconnected; will be updated when BlueprintMachine is implemented
+    //std::string resName = std::get<std::string>(settings.m_config["resourcename"]);
+    std::string_view resName = "";
+    Path resPath = decompose_path(resName);
+    Package& pkg = m_scene.get_application().debug_find_package(resPath.prefix);
+
+    ShipResource resource;
+    resource.m_type = pkg.get<ShipResourceType>(resPath.identifier);
+    resource.m_quantity = resource_capacity(*resource.m_type, capacity);
+
+    return m_scene.reg_emplace<MachineContainer>(ent, capacity, resource);
```

Comment:
This function will change completely in the next PR when everything is hooked up; for now I just have it this way so that the file compiles.